core/domain/translation_domain.py

```diff
@@ -18,9 +18,262 @@
 
 from __future__ import annotations
 
+import enum
+
+from core import feconf
 from core import utils
 
-from typing import Dict
+from typing import Dict, List
+from typing_extensions import TypedDict
+
+
+class TranslatableContentFormat(enum.Enum):
+    """Represents all possible data types for any translatable content."""
+
+    HTML = 'html'
+    UNICODE_STRING = 'unicode'
+    SET_OF_NORMALIZED_STRING = 'set_of_normalized_string'
+    SET_OF_UNICODE_STRING = 'set_of_unicode_string'
+
+
+class TranslatableContentDict(TypedDict):
+    """Dictionary representing TranslatableContent object."""
+
+    content_id: str
+    content: feconf.ContentInTranslatableContent
+    content_type: TranslatableContentFormat
+
+
+class TranslatableContentsCollection:
+    """A class to collect all TranslatableContents from a translatable object
+    and maps with their corresponding content-ids.
+    """
+
+    def __init__(self) -> None:
+        """Constructs a TranslatableContentsCollection object."""
+        self.translatable_contents: Dict[str, TranslatableContent] = {}
+
+    def add_translatable_field(
+        self,
+        field_type: TranslatableContentFormat,
+        content_id: str,
+        value: feconf.ContentInTranslatableContent
+    ) -> None:
+        """Adds translatable field parameter to translatable_contents dict.
+
+        Args:
+            field_type: TranslatableContentFormat. The type of the
+                corresponding translatable content.
+            content_id: str. The id of the corresponding translatable content.
+            value: ContentInTranslatableContent. Value of the content which is
+                translatable.
+        """
+        self.translatable_contents[content_id] = TranslatableContent(
+            content_id, value, field_type)
+
+    def add_translatable_object(
```

Comment:
add_fields_from_translatable_object

Though I'm wondering about the framing here. Isn't this basically a merge? If so, then maybe have this be extend(other_translatable_contents_collection) if there isn't expected to be duplication between the passed-in collection and the existing one; otherwise have this be update(other_translatable_contents_collection) to follow regular Python dict semantics. Either way, in both cases, pass in a translatable_contents_collection and not some external object.

Also: in the first case (extend, no duplicates), what happens if we see a conflict, do/should we check for that?

Reply:
> Updated based on discussion.

As per offline discussion, we finalized to have `add_fields_from_translatable_object` with `translatable_object` as param because of the following reasons:
- The logic to fetch translation_collections from translatable_objects is similar so the calling site for the `extend` method will have duplicate logic everywhere.
-  `add_fields_from_translatable_object` having the general logic to generate entire collections out of translation objects will make it easy for maintenance and will leave the calling site clean and redable.

Reply:
@Nik-09 For (2): the link you provided is to a comment I had with 4 questions, none of which were actually answered. So I still don't know the outcome or what the thought process is here, and my question still stands.

Reply:
> 1. Do we ever delete items from the registration dict? If not, then how do you handle the "hint gets deleted" case?

Hint gets deleted case is just an example where I want to show the states of an object changes in its lifecycle.
If any of the hints from the hint_list gets deleted then the successive calling of get_translatable_contents_collection() (based on new architecture) will return the updated hint list and get stored in the dict content_id_to_translatable_content.
This phenomenon to get the updated state of translatable content is similar for every translatable object.

> 2. I don't follow the argument about duplicates. In a composite structure, why would we have duplicates in the first place -- wouldn't that already indicate that something is wrong? What's a legitimate case in which we have duplicates?

After a discussion with @DubeySandeep, we came to the conclusion that duplicate checking is not necessary. Hence it is removed.

> 3. "In future if we add any other functionality to BaseTranslatableObject then it would be easy to access all translatable contents via _translatable_contents." -- The problem with this is that it means you have to automatically keep _translatable_contents up-to-date any time a change is made to the object. How are you planning to do that? The two possible ways are to (a) recompute on each get, or (b) update the inner field on each update; I think you're going with (a), but that means that translatable content should always be accessed via the public get() method -- is that right? In that case I think it's OK to keep the dict as local within the get() method so that it doesn't get used outside its intended context.

 Updated the architecture, thus no variable at the class level of BaseTranslatableObject.
Initially BaseTranslatableObject class was providing both storage and service functionality so because of that there was a moment in which a class-level variable data becomes inconsistent and for every get call we have to compute the value from scratch.
Now we have separated the architecture into different classes
TranslatableContentsCollection (for storage)
BaseTranslatableObject (for service)
This solves the problem of storing inconsistent data at the class level, so new architecture do not need local variable too because the method (get_translatable_field()) which is concerned here is removed.

> 4. How many _times_ does _get_all_translatable_content() get called within a single request/response (typically)? Is it once for each of the bullet points you mentioned?

Yes, it is called once for each method that was mentioned in the bullet.

Reply:
The new architecture was designed in order to align with all the suggestions that were mentioned in that thread.
Although I have added responses to each of the questions, please take a look.
Also, I am open to answering follow-ups.

Reply:
Thanks @Nik-09. I am reproducing the list of bullet points here since the original comment is hard to find:

- get_translation_counts()
- get_translatable_content_count()
- has_content_id()
- get_content_id_mapping_needing_translations()

That sounds like 4 calls per request. In each of these calls it needs to be recomputed from scratch, correct?

Also, I just have one follow-up question. It is the one about duplicates: if we see a conflict, shouldn't we check for that and raise if it happens? I don't yet see a good explanation in this thread either for or against. I would lean towards raising just so that, if some unexpected issue happens, we don't let it slip by and can do something about it, otherwise (if I'm not mistaken) it'll be overwritten silently.

Reply:
I'm replying to the above question so the we can resolve the thread quickly (@Nik-09 Feel free to add a response if I have missed anything)

> if we see a conflict, shouldn't we check for that and raise if it happens?

The entity validation check will ensure that the content_id is unique throughout the exploration. So other API/methods relies on that (we follow this structure for all other methods in the domain layer.) + We have the policy to make entity.validation() call before saving models to datastore.

> In each of these calls it needs to be recomputed from scratch, correct?

These function calls are happening on different requests and some requests will make such multiple computations in one request but in taskqueue.

Reply:
The validation thing is fine but please add a comment in the code to explain the reasoning since it is a key assumption that we need to make sure always holds. Also perhaps anywhere we have such a check, we need a comment there saying "don't remove this, this structure depends on it". That should be enough to keep it foolproof, I think.

For the function calls: sounds good, does that mean that one non-taskqueue request will make at most one such call?

Reply:
Yeah, we will add a code comment: `The duplicate check is avoided here as object validation check ensures all the content_ids are unique.`

> Also perhaps anywhere we have such a check, we need a comment there saying "don't remove this, this structure depends on it".

We are not adding any `if` check here.

> For the function calls: sounds good, does that mean that one non-taskqueue request will make at most one such call?

Yes, it will make one such call.

Reply:
Suggest: "We aren't checking for duplicate values here, since the validate() check for each object will ensure that all the content_ids are unique." (I.e. being more explicit about what the "object validation check" is, and providing more context for the "duplicate check".)

Also, I don't understand what you mean by the "if" check bit. What I'm saying is, for parts of the validate() function that refer to content IDs, a comment should be added to explain not to remove or weaken it.

Reply:
*Added comment in add_translatable_field() method.

Reply:
> Ah, I see! The [duplicate content_id validation check](https://github.com/oppia/oppia/blob/6d59af89be3cadbdc6702e4149fe616adf3bd890/core/domain/state_domain.py#L2548) for content ids existed in the object validation method for long, we just have to refactor those checks while adding this new structure to the objects. I'll make sure to add the comment but I also think we don't expect to remove any check from the domain validation method anytime without any proper explanation/reasoning.

Sounds good -- yup, confirming that we should add it. This is because there is a dependency here that should be made explicit, and future developers may not know about the link if it's not clearly called out.

Reply:
@seanlip We have added an issue (https://github.com/oppia/oppia/issues/15071) to track the task to add code comments while refactoring the validation check.

Also, I think we haven't added such comment in the codebase and most of the written_translation code are written with an assumption of content_id to be unique and being validated by the validation method.